update README and docs to current project status (#118)
NetherlandsForensicInstitute/asmtransformers: asmtransformers/docs/architecture.md

## Purpose

`asmtransformers` is the training and inference core of the monorepo. It turns control-flow-graph representations of 
assembly functions into token sequences, feeds those sequences into transformer models, and exposes a script-oriented 
workflow for preprocessing, pretraining, finetuning, evaluation, and embedding generation.

## Main Subsystems

### ISA preprocessing

The ISA-specific preprocessing path is implemented in [asmtransformers.preprocessors](../asmtransformers/preprocessors/__init__.py), with current implementations 

> for ARM64, x86/amd64, and RISC-V. I386 is also supported, but gets preprocessed using the x86 preprocessor.

Reviewer: Think it'd be nice to keep to the same values we're also using in the code

- amd64
- arm64
- i386
- riscv64

(also: x86 is more i386 than it is amd64, which is often also referred to as x86_64)

```suggestion
for amd64, arm64, i386 and riscv64 (though amd64 and i386 are preprocessed by the same `X86Preprocessor`).
```

### ISA preprocessing

> TO DO: i386 is processed by which preprocessor?

Reviewer: The `X86Preprocessor`, see `ASMTokenizer.__init__`.

These helpers reduce token explosion caused by raw numeric values. The current tokenizer setup uses:

> TO DO: is this still true? I'm not sure we call both anymore

Reviewer: Yes. Those formatters are included by default, see `ASMTokenizer.__init__`.

The main layers are:

TO DO: I believe the below is up to date but let's check

- `ASMBertForMaskedLM` and `ASMBertModel` adapt Hugging Face BERT classes to the jTrans-style setup, including shared word/position embeddings and jump-target prediction support during pretraining.

> - `build_finetuning_model()` adapts the pretrained transformer into a plain `SentenceTransformer` model for triplet-loss finetuning.

Reviewer 2: This function does not exist anymore

Author: it does, but it lives in asmsentencebert.py. I'm trying to figure out if what's written here does the current code justice or whether it should be adapted

Reviewer 2: Oh My bad! I see it is also still being used in `finetune.py`, i missed it completely. Sorry!

- packaged model assets in `models/arm64bert/`

> - scripts and runtime paths that default to `arm64` when no architecture is supplied

Reviewer: Ah, this is where that line was; this is no longer the case, defaults have been removed everywhere, architecture needs to specified explicitly.

Author: thanks for telling me! that makes this section kinda redundant, we already describe in the ISA preprocessing section  where to see the ARM specific preprocessing. DO you mind if I remove this section?

Reviewer: If that was the sole reason for it, ofc not, I DO not mind :)

TO DO: is the second line still correct?

Preprocessing itself is no longer ARM64-only: `ASMTokenizer` also dispatches to x86/amd64 and RISC-V preprocessors.

> The multilingual model can be found on [Huggingface](https://huggingface.co/NetherlandsForensicInstitute/Multilingual-ASMBERT)

Reviewer: Hugging Face uses two capitalized words in their own header

```suggestion
The multilingual model can be found on [Hugging Face](https://huggingface.co/NetherlandsForensicInstitute/Multilingual-ASMBERT)
```

## What Is Reusable For Other ISAs Today

TO DO: is this correct? did I miss anything?

If you want to create your own model for a different assembly architecture using this repo, we recommend you take the

> following steps:

Reviewer: These steps all refer to files whose paths are neither rooted at the repository nor at the python module; why strip `asmtransformers/` if we're trying to point someone to a specific file? Other file references in this very file use the full relative path with a link to that file. Maybe we ought to pick one of the options, both seems confusing atm.

Author: better?

Reviewer: Yeah think so, it's all the same now, in any case 🙏🏻

- create an architecture-specific preprocessor in `models/preprocessors`

> - ensure the architecture name is specified in `models/__init__.py`

Reviewer: ```suggestion
- ensure the architecture name is specified in the `Architecture` enumeration in `__init__.py`
```

> - ensure it is specified in `ASMTokenizer` in `models/multilingual_asmbert/asmbert.py`

Reviewer: That file does not exist

```suggestion
- ensure it is included in `ASMTokenizer` in `models/asmbert.py`
```

- packaged model assets in `models/arm64bert/`
- scripts and runtime paths that default to `arm64` when no architecture is supplied

> TO DO: is the second line still correct?

Reviewer: Until we release the new one on huggingface, I guess?

Author: I'll paste the link already, assuming that we will release it this week.

Preprocessing itself is no longer ARM64-only: `ASMTokenizer` also dispatches to x86/amd64 and RISC-V preprocessors.
There is currently no packaged vocabulary or trained model for non-ARM64 ISAs.

## What Is Reusable For Other ISAs Today

> TO DO: is this section still relevant?

Reviewer: Good point, the current state of the code base proves that the bits are reusable, I'd say :)

Author: I'm doubting between
* keeping it so that people who want to use it for other ISA's/intermediate representations/other stuff can do so
* just removing the section bc we covered the biggest things
What do you think?

Reviewer: I don't know who's going to read this. The *code* is clearly capable of handling different ISA's, that's both demo and documentation. If it's useful to include in the readme here, maybe some pointers to where this is implemented might help?

Author: I changed it to a section about what people should do if they want to train a model with a different ISA, can you check if it makes sense?
